hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SNDArray.scala

```diff
@@ -180,24 +185,134 @@ trait SNDArray extends SType {
+    }
+    outOfBounds
+  }
 
-  def sameShape(other: SNDArrayValue, cb: EmitCodeBuilder): Code[Boolean]
+  def assertInBounds(indices: IndexedSeq[Value[Long]], cb: EmitCodeBuilder, errorId: Int): Code[Unit] = {
+    val shape = this.shapes(cb)
+    Code.foreach(0 until st.nDims) { dimIndex =>
+      val eMsg = const("Index ").concat(indices(dimIndex).toS)
+        .concat(s" is out of bounds for axis $dimIndex with size ")
+        .concat(shape(dimIndex).toS)
+      (indices(dimIndex) >= shape(dimIndex)).orEmpty(Code._fatalWithID[Unit](eMsg, errorId))
+    }
+  }
 
-  def firstDataAddress(cb: EmitCodeBuilder): Value[Long]
+  def sameShape(other: SNDArrayValue, cb: EmitCodeBuilder): Code[Boolean] = {
+    val otherShapes = other.shapes(cb)
+    val b = cb.newLocal[Boolean]("sameShape_b", true)
+    val shape = this.shapes(cb)
+    assert(shape.length == otherShapes.length)
+    shape.zip(otherShapes).foreach { case (s1, s2) =>
+      cb.assign(b, b && s1.ceq(s2))
+    }
+    b
+  }
+
+  def slice(cb: EmitCodeBuilder, indices: IndexedSeq[NDArrayIndex]): SNDArraySliceCode = {
+    val shapeX = shapes(cb)
+    val stridesX = strides(cb)
+    val shapeBuilder = mutable.ArrayBuilder.make[Code[Long]]
+    val stridesBuilder = mutable.ArrayBuilder.make[Code[Long]]
+
+    for (i <- indices.indices) indices(i) match {
+      case ScalarIndex(j) =>
+        cb.ifx(j < 0 || j >= shapeX(i), cb._fatal("Index out of bounds"))
+      case SliceIndex(Some(begin), Some(end)) =>
+        cb.ifx(begin < 0 || end > shapeX(i) || begin > end, cb._fatal("Index out of bounds"))
+        shapeBuilder += end - begin
+        stridesBuilder += stridesX(i)
+      case SliceIndex(None, Some(end)) =>
+        cb.ifx(end >= shapeX(i) || end < 0, cb._fatal("Index out of bounds"))
+        shapeBuilder += end
+        stridesBuilder += stridesX(i)
+      case SliceIndex(Some(begin), None) =>
+        val end = shapeX(i)
+        cb.ifx(begin < 0 || begin > end, cb._fatal("Index out of bounds"))
+        shapeBuilder += end - begin
+        stridesBuilder += stridesX(i)
+      case SliceIndex(None, None) =>
+        shapeBuilder += shapeX(i)
+        stridesBuilder += stridesX(i)
+      case ColonIndex =>
+        shapeBuilder += shapeX(i)
+        stridesBuilder += stridesX(i)
+    }
+    val newShape = shapeBuilder.result()
+    val newStrides = stridesBuilder.result()
+
+    val firstElementIndices = indices.map {
+      case ScalarIndex(j) => j
+      case SliceIndex(Some(begin), _) => begin
+      case SliceIndex(None, _) => const(0L)
+      case ColonIndex => const(0L)
+    }
+
+    val newFirstDataAddress = loadElementAddress(firstElementIndices, cb)
+
+    val newSType = SNDArraySlice(PCanonicalNDArray(st.pType.elementType, newShape.size, st.pType.required))
+
+    new SNDArraySliceCode(newSType, newShape, newStrides, newFirstDataAddress)
+  }
+
+  def slice(cb: EmitCodeBuilder, indices: Any*): SNDArraySliceCode = {
```

Comment:
I'm not a huge fan of this helper function -- we don't need the syntactic sugar in the compiler. I weight clarity highly above terseness in the backend, and I think it'll be easier to maintain the single slice function
```
  def slice(cb: EmitCodeBuilder, indices: IndexedSeq[NDArrayIndex]): SNDArraySliceCode = {
```

Reply:
I definitely understand the concern. But a method like local whitening that's implemented using our codegen abstractions blurs the line between "in the compiler" and out. It's a complicated method that's already difficult to read and reason about in Julia, and will much harder to follow in Hail even with sugar like this.

I can drop it from this PR, and we can weigh the benefits later with some before/after code samples from the local whitening implementation.

Reply:
OK, sounds good re: looking at the code samples with local whitening. And I definitely do see the need for ergonomics in the compiler!